src/shims/time.rs

```diff
@@ -107,6 +109,93 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
         Ok(0)
     }
 
+    // The localtime() function shall convert the time in seconds since the Epoch pointed to by
+    // timer into a broken-down time, expressed as a local time.
+    // https://linux.die.net/man/3/localtime_r
+    #[allow(clippy::arithmetic_side_effects)]
+    fn localtime_r(
+        &mut self,
+        timep: &OpTy<'tcx, Provenance>,
+        result_op: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, Pointer<Option<Provenance>>> {
+        let this = self.eval_context_mut();
+
+        this.assert_target_os_is_unix("localtime_r");
+        this.check_no_isolation("`localtime_r`")?;
+
+        let result = this.deref_pointer_as(result_op, this.libc_ty_layout("tm"))?;
+        let result_ptr = this.read_pointer(result_op)?;
+
+        let sec_since_epoch_signed: i64 =
+            this.read_scalar(&this.deref_pointer(timep)?)?.to_i64()?;
+        let sec_since_epoch: u64 = sec_since_epoch_signed.try_into().unwrap();
+        let new_time = UNIX_EPOCH + Duration::from_secs(sec_since_epoch);
+
+        let dt: DateTime<Local> = new_time.into();
+
+        let tm_sec = dt.second();
+        let tm_min = dt.minute();
+        let tm_hour = dt.hour();
+        let tm_mday = dt.day();
+        let tm_mon = dt.month0();
+        let tm_year = dt.year() - 1900;
+        let tm_wday = dt.weekday().num_days_from_sunday();
+        let tm_yday = dt.ordinal0();
+
+        // This value is always set to -1, because there is no way to know if dst is in effect with
+        // chrono crate yet
+        // This may not be consistent with libc::localtime_r's result
+        let tm_isdst = -1;
+
+        // tm_zone represents the timezone value in the form of: +0730, +08, -0730 or -08
+        // This may not be consistent with libc::localtime_r's result
+        let offset_in_second = Local::now().offset().local_minus_utc();
+        let tm_gmtoff = offset_in_second;
+        let mut tm_zone_str = String::new();
+        if offset_in_second < 0 {
+            tm_zone_str.push('-');
+        } else {
+            tm_zone_str.push('+');
+        }
+        let offset_hour = format!("{:0>2}", (offset_in_second.abs() / 3600).to_string());
+        tm_zone_str.push_str(&offset_hour);
+        let offset_min = (offset_in_second % 3600) / 60;
+        if offset_min != 0 {
+            tm_zone_str.push_str(&offset_min.to_string());
+        }
+
+        // Allocate pointer for tm_zone string
+        let byte_str = tm_zone_str.as_bytes();
+        let size: u64 = byte_str.len().try_into().unwrap();
+        let kind = MiriMemoryKind::C;
+        let ptr = this
+            .allocate_ptr(Size::from_bytes(size + 1), Align::from_bytes(2).unwrap(), kind.into())
+            .unwrap();
```

Comment:
This is your memory leak. You want to use `MiriMemoryKind::Machine`. Also there's a convenience helper called `allocate_str` that does the whole "allocate then write" dance for you

Reply:
How is the memory for the time zone string managed in the real implementation?

We shouldn't allocate a new timezone string that is then lekaed each time this function is called. So if this becomes `Machine` memory then it should be allocated once at program startup.

Reply:
This is what we do for `ERRNO` and panic messages, too. If we want to deduplicate things, we should add a general mechanism. It's not something really new from this PR.

https://www.man7.org/linux/man-pages/man3/tm.3type.html says 

> tm_zone points to static storage and may be overridden on
       subsequent calls to [localtime(3)](https://www.man7.org/linux/man-pages/man3/localtime.3.html) and similar functions (however,
       this never happens under glibc).

it also says

> tm_zone is the equivalent of [tzname(3)](https://www.man7.org/linux/man-pages/man3/tzname.3.html) for the timezone
       represented by this broken-down time.

So we could teach miri about `tmzone` and always fetch it to use its data, but it's not really clear to me what the right behaviour here is

Most `tm` man pages don't even contain this extra field at all.

Reply:
> This is what we do for ERRNO and panic messages, too. 

Not for errno, those are written to a caller-provided buffer.

But for panic messages, indeed. We should probably make this stack-allocated, the panic handler will make a copy of this anyway so the panic message should be freed when the panic handler returns. Or we add a de-duplicating "interner" that makes sure we allocate each string only once (i.e., this acts as-if it was a static string embedded in the binary).

> may be overridden on subsequent calls  to [localtime(3)](https://www.man7.org/linux/man-pages/man3/localtime.3.html)

But crucially, not for `localtime_r` as that would not be thread-safe. Would be interesting to figure out what glibc does here. Probably the parse the timezone info into a single global variable once and then just return pointers to that.

I'm okay with leaving this as a FIXME, and opening an issue to do string de-duplication.

Reply:
> It seems like they malloc it, but I didn't see it being freed anywhere yet.

They malloc it once and then keep a global cache.

> This issue is tracked under https://github.com/rust-lang/miri/issues/3470.

Thanks! That works for now.